### ISSUE TITLE
Labelled state/reader variables

### ISSUE THREAD

**Issue by robrix** [fused-effects/fused-effects#264]
`Reader` and `State` both suffer from a couple of inconveniences:

- operations like `get` and `ask` often need to be disambiguated with e.g. `TypeApplications` because of the lack of injectivity—_which_ variable are you `get`ting?

- since `Reader`/`State` variables are distinguished solely by type, `newtype`s are required to avoid accidentally shadowing variables of the same type. (This can have readability benefits, but can also be inconvenient when using code which didn’t have the foresight to do so.)

We could ameliorate both of these problems by means of labels, à la `capabilities`:

```haskell
labelLength :: HasLabelled "label" (Reader [a]) sig m => m Int
labelLength = gets @"label" length
```

Better yet, by making labels poly-kinded, we can allow specific type tags to be used, allowing for usage without accidental capture (as is possible with stringly-typed selection) _or_ needless unwrapping:

```haskell
data Label

labelLength :: HasLabelled Label (Reader [a]) sig m => m Int
labelLength = gets @Label length
```

This works because the requirement of the label allows us to make the selection of the effect injective: there can be only one (visible) effect for a given label.

I’ve [prototyped this out](https://github.com/fused-effects/diffused-effects/pull/4) in `diffused-effects`, and it works pretty nicely; it also doesn’t require that we alter any of the existing machinery, so it could be easily deployed as a separate library.

**Comment by patrickt** [comment]
I’d be in favor of a `fused-effects-labeled` library. I don’t think the value proposition is quite high enough to merit inclusion in the core, but I know a lot of people use and like this style, so we should give them the option.

**Comment by robrix** [comment]
We shipped this in 1.0.2.
